### PR TITLE
Fix TalkBack incorrectly reading out spans on StoryPromo content

### LINKED ISSUE #9561: TalkBack incorrectly reading links in nested spans
**Describe the bug**
When using TalkBack on Android, links with role text and nested spans aren't read as one item as intended to be. Content is also repeated in additional blocks/gestures.

**To Reproduce**
Steps to reproduce the behaviour:

1. Go to https://www.bbc.com/portuguese/podcasts/p07r3r3t
2. Navigate to the 'O podcast está disponível em' header on TalkBack
3. Swipe to the list on TalkBack
4. TalkBack should read 'Spotify, Que Historia, externo, link'
5. Swipe through the links and TalkBack should read an additional 'Spotify', and the text highlighted
6. Swipe again and TalkBack will read out 'Que Historia' as well as highlight a separate 'block' on the screen.

**Expected behaviour**
TakBack should only read out 'Spotify, Que Historia, externo, link' before moving on to the next link.

**Screenshots**
image

**Smartphone (please complete the following information):**

- Device: Samsung S10
- OS: Android 11
- Browser: Chrome
- Version: 94.0.4606.85

**Testing notes**

- [x] This bug fix is expected to need manual testing.

**Checklist**
- [x] (BBC contributors only) This issue follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Additional context**

### PR DESCRIPTION
**Pull request by nicola-d-bbc**
Resolves #9561

**Overall change:**
Adds unique ID to nested span to allow link to use aria-labelledby.

**Code changes:**

- Adds aria-labelledby to link of storyPromoLinkText{mediaID}, using unique media ID from item content
- Adds ID to span with role="text" to allow nested spans to be treated as one gesture in TalkBack

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [x] I have assigned this PR to the Simorgh project
- [x] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**

- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/containers/CpsFeaturesAnalysis/__snapshots__/index.test.jsx.snap [hunk 1]
```diff
@@ -696,6 +696,7 @@ exports[`CpsRelatedContent should render Story Feature components when given app
                 class="emotion-37 emotion-38"
               >
                 <a
+                  aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:igbo/afirika-23252735"
```
Comment (greenc05) [inline src/app/containers/CpsFeaturesAnalysis/__snapshots__/index.test.jsx.snap]: Should we have aria-labelledby in this snapshot? I don't see any nested spans?

File: src/app/containers/CpsRelatedContent/RelatedContentPromo/__snapshots__/index.test.jsx.snap [hunk 2]
```diff
@@ -351,10 +351,12 @@ exports[`RelatedContentPromo it renders a Story Promo wrapped in a Grid componen
           class="emotion-14 emotion-15"
         >
           <a
+            aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:pidgin/44508901"
```
Comment (greenc05) [inline src/app/containers/CpsRelatedContent/RelatedContentPromo/__snapshots__/index.test.jsx.snap]: Should we have aria-labelledby here? I don't see more than 1 nested span with content?

File: src/app/containers/CpsRelatedContent/RelatedContentPromoList/__snapshots__/index.test.jsx.snap [hunk 3]
```diff
@@ -427,10 +427,12 @@ exports[`RelatedContentPromoList it renders a list of Story Promos for MAP pages
             class="emotion-19 emotion-20"
           >
             <a
+              aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:pidgin/44508901"
```
Comment (greenc05) [inline src/app/containers/CpsRelatedContent/RelatedContentPromoList/__snapshots__/index.test.jsx.snap]: Same, not sure we need aria-labelledby in this snapshot?

File: src/app/containers/CpsRelatedContent/__snapshots__/index.test.jsx.snap [hunk 4]
```diff
@@ -805,10 +805,12 @@ exports[`CpsRelatedContent should render Story Promo components when given appro
                 class="emotion-39 emotion-40"
               >
                 <a
+                  aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:pidgin/44508901"
```
Comment (greenc05) [inline src/app/containers/CpsRelatedContent/__snapshots__/index.test.jsx.snap]: Same...

File: src/app/containers/CpsTopStories/__snapshots__/index.test.jsx.snap [hunk 5]
```diff
@@ -615,6 +615,7 @@ exports[`CpsRelatedContent should render Top Stories components when given appro
                 class="emotion-29 emotion-30"
               >
                 <a
+                  aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:mundo/noticias-internacional-51939501"
```
Comment (greenc05) [inline src/app/containers/CpsTopStories/__snapshots__/index.test.jsx.snap]: And same comment...

File: src/app/containers/FrontPageStoryRows/__snapshots__/index.test.jsx.snap [hunk 6]
```diff
@@ -880,6 +880,7 @@ exports[`FrontPageStoryRows Container - snapshots LeadingRow - rtl 1`] = `
           class="emotion-46 emotion-47"
         >
           <a
+            aria-labelledby="storyPromoLinkText-cd083f06-0475-4f92-b42b-bc4c54eb22ea"
```
Comment (greenc05) [inline src/app/containers/FrontPageStoryRows/__snapshots__/index.test.jsx.snap]: Not needed in this snapshot either?

File: src/app/containers/IndexPageSection/__snapshots__/index.test.jsx.snap [hunk 7]
```diff
@@ -1018,6 +1018,7 @@ HTMLCollection [
                 class="emotion-37 emotion-38"
               >
                 <a
+                  aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:igbo/testasset-00000001"
```
Comment (greenc05) [inline src/app/containers/IndexPageSection/__snapshots__/index.test.jsx.snap]: Or this one?

File: src/app/containers/StoryPromo/__snapshots__/index.test.jsx.snap [hunk 8]
```diff
@@ -3664,6 +3680,7 @@ exports[`StoryPromo Container should render featureLink correctly for amp 1`] =
         class="emotion-12 emotion-13"
       >
         <a
+          aria-labelledby="storyPromoLinkText-undefined"
```
Comment (greenc05) [inline src/app/containers/StoryPromo/__snapshots__/index.test.jsx.snap]: Other examples above are good, not need here though?

Comment (greenc05) [reply src/app/containers/StoryPromo/__snapshots__/index.test.jsx.snap]: Or in several places below this?

File: src/app/pages/IdxPage/__snapshots__/index.test.jsx.snap [hunk 9]
```diff
@@ -2420,6 +2420,7 @@ exports[`IdxPage - Persian Snapshots should render a persian idx page correctly
                 class="emotion-37 emotion-38"
               >
                 <a
+                  aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:persian/afghanistan-52735886"
```
Comment (greenc05) [inline src/app/pages/IdxPage/__snapshots__/index.test.jsx.snap]: Not needed here?

File: src/app/pages/IdxPage/__snapshots__/index.test.jsx.snap [hunk 10]
```diff
@@ -3663,6 +3674,7 @@ exports[`IdxPage - Persian Snapshots should render a persian idx page correctly
                   class="emotion-37 emotion-38"
                 >
                   <a
+                    aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:persian/afghanistan-52722865"
```
Comment (greenc05) [inline src/app/pages/IdxPage/__snapshots__/index.test.jsx.snap]: Not needed here? Or in the links below?

File: src/app/pages/MediaAssetPage/__snapshots__/index.test.jsx.snap [hunk 11]
```diff
@@ -4363,6 +4363,7 @@ exports[`Media Asset Page should render component 1`] = `
                           class="emotion-264 emotion-265"
                         >
                           <a
+                            aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:pidgin/tori-23145776"
```
Comment (greenc05) [inline src/app/pages/MediaAssetPage/__snapshots__/index.test.jsx.snap]: Can we check this snapshot too please?

File: src/app/pages/PhotoGalleryPage/__snapshots__/index.test.jsx.snap [hunk 12]
```diff
@@ -6351,10 +6351,12 @@ exports[`Photo Gallery Page snapshots should match snapshot for PGL with all CPS
                           class="emotion-235 emotion-236"
                         >
                           <a
+                            aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:pidgin/media-45132087"
                             class="emotion-237 emotion-238"
                             href="/pidgin/media-45132087"
                           >
                             <span
+                              id="storyPromoLinkText-urn:bbc:ares::asset:pidgin/media-45132087"
```
Comment (greenc05) [inline src/app/pages/PhotoGalleryPage/__snapshots__/index.test.jsx.snap]: Not needed here?

File: src/app/pages/PhotoGalleryPage/__snapshots__/index.test.jsx.snap [hunk 13]
```diff
@@ -11287,6 +11291,7 @@ exports[`Photo Gallery Page snapshots should match snapshot for PGL with non-CPS
                           class="emotion-449 emotion-450"
                         >
                           <a
+                            aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:azeri/azerbaijan-44124647"
```
Comment (greenc05) [inline src/app/pages/PhotoGalleryPage/__snapshots__/index.test.jsx.snap]: Not needed here or in the links below?

File: src/app/pages/StoryPage/__snapshots__/index.test.jsx.snap [hunk 14]
```diff
@@ -3014,6 +3014,7 @@ exports[`Story Page should not show the pop-out timestamp when allowDateStamp is
                               class="emotion-211 emotion-212"
                             >
                               <a
+                                aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:igbo/23182501"
```
Comment (greenc05) [inline src/app/pages/StoryPage/__snapshots__/index.test.jsx.snap]: Not needed here?

File: src/app/pages/StoryPage/__snapshots__/index.test.jsx.snap [hunk 15]
```diff
@@ -3215,6 +3219,7 @@ exports[`Story Page should not show the pop-out timestamp when allowDateStamp is
                               class="emotion-284 emotion-212"
                             >
                               <a
+                                aria-labelledby="storyPromoLinkText-urn:bbc:ares::asset:igbo/23182930"
```
Comment (greenc05) [inline src/app/pages/StoryPage/__snapshots__/index.test.jsx.snap]: And again and several links below?

**Comment by greenc05** [comment]
@nicola-d-bbc apologies for all the comments, is it not possible to apply this if it has media or live? Rather than blanket apply it everywhere?

File: src/app/containers/StoryPromo/index.jsx [hunk 16]
```diff
@@ -178,6 +188,7 @@ const StoryPromoContainer = ({
 
   const Info = (
     <>
+      {console.log(labelId === 'rel-content' ? item : '')}
```
Comment (jroebu14) [inline src/app/containers/StoryPromo/index.jsx]: console.log needing removed

**Comment by DarioR01** [comment]
> @nicola-d-bbc apologies for all the comments, is it not possible to apply this if it has media or live? Rather than blanket apply it everywhere?

Yes, I will try to take care of some of them. But I don't think they can all be taken away

File: src/app/containers/StoryPromo/index.jsx [hunk 17]
```diff
@@ -40,6 +40,11 @@ const SingleColumnStoryPromo = styled(StoryPromo)`
   }
 `;
 
+const getLinkId = (assetUri, labelId) => {
+  const assetId = assetUri.split('/').pop();
+  return `promo-link-${labelId}${assetId || ''}`;
```
Comment (ryanmccombe) [inline src/app/containers/StoryPromo/index.jsx]: Can we make sure IDs only include valid characters?  eg a-z, A-Z and 0-9

**Review by ryanmccombe — APPROVED** [review]
👍

File: src/reports/jest/junit-1d2eb740-4633-11ec-8620-4ff8a6d7712d.xml [hunk 18]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jroebu14) [inline src/reports/jest/junit-1d2eb740-4633-11ec-8620-4ff8a6d7712d.xml]: not sure this should be committed

Comment (DarioR01) [reply src/reports/jest/junit-1d2eb740-4633-11ec-8620-4ff8a6d7712d.xml]: No, they should not, deleting them now.

**Review by jroebu14 — APPROVED** [review]
Nice one!

**Pull request merged**